plugins/BEdita/API/src/Auth/EndpointAuthorize.php

```diff
@@ -0,0 +1,200 @@
+
+use BEdita\Core\Model\Entity\Application;
+use BEdita\Core\Model\Entity\Endpoint;
+use BEdita\Core\Model\Entity\EndpointPermission;
+use BEdita\Core\State\CurrentApplication;
+use Cake\Auth\BaseAuthorize;
+use Cake\Datasource\Exception\RecordNotFoundException;
+use Cake\Http\ServerRequest;
+use Cake\Network\Exception\ForbiddenException;
+use Cake\ORM\TableRegistry;
+use Cake\Utility\Hash;
+
+/**
+ * Provide authorization on a per-endpoint basis.
+ *
+ * @since 4.0.0
+ */
+class EndpointAuthorize extends BaseAuthorize
+{
+
+    /**
+     * {@inheritDoc}
+     */
+    protected $_defaultConfig = [
+        'apiKeyHeaderName' => 'X-Api-Key',
+    ];
+
+    /**
+     * Current endpoint entity.
+     *
+     * @var \BEdita\Core\Model\Entity\Endpoint|null
+     */
+    protected $endpoint = null;
+
+    /**
+     * Request object instance.
+     *
+     * @var \Cake\Http\ServerRequest
+     */
+    protected $request;
+
+    /**
+     * Cache result of `authorized()` method call.
+     *
+     * This is required for controller to know whether authorization was granted on all contents,
+     * or only on those that belong to the current user. Whatever that means, it is controller's
+     * responsibility to interpret, as it may vary. Some controller may also decide to ignore this
+     * fine-grained authorization level.
+     *
+     * @var bool|string
+     */
+    protected $authorized;
+
+    /**
+     * {@inheritDoc}
+     */
+    public function authorize($user, ServerRequest $request)
+    {
+        $this->request = $request;
+
+        $application = $this->getApplication();
+        $endpoint = $this->getEndpoint();
+        $permissions = $this->getPermissions($user, $application, $endpoint)->toArray();
+
+        $this->authorized = $this->checkPermissions($permissions);
+
+        if (!empty($user['_anonymous']) && $this->authorized !== true) {
+            // Anonymous user should not get a 403. Thus, we invoke authentication provider's
+            // `unauthenticated()` method. Furthermore, for anonymous users, `mine` doesn't make any sense,
+            // so we treat that as a non-authorized request.
+            $controller = $this->_registry->getController();
+
+            $controller
+                ->Auth->getAuthenticate('BEdita/API.Jwt')
+                ->unauthenticated($controller->request, $controller->response);
+        }
+
+        // Authorization is granted for both `true` and `'mine'` values.
+        return !empty($this->authorized);
+    }
+
+    /**
+     * Get application for request.
+     *
+     * @return \BEdita\Core\Model\Entity\Application|null
+     * @throws \Cake\Network\Exception\ForbiddenException Throws an exception if API key is missing or invalid.
+     */
+    protected function getApplication()
+    {
+        $application = CurrentApplication::getApplication();
+        if ($application === null) {
+            try {
+                CurrentApplication::setFromApiKey(
```

Comment:
API Keys are not an optional feature, but in many cases like development, test environment or initial setups you may choose not to use them.

Maybe with a simple config like ```['ignoreApikeys' => true]``` or an empty `applications` table we could create a default application at startup. Since we have to remove api keys from config, we may define the expected behavior as we work on that.

Ok for now